src/chainstate/stacks/index/file.rs

```diff
@@ -329,60 +652,83 @@ impl TrieFile {
 
-impl NodeHashReader for TrieFileNodeHashReader<'_> {
-    fn read_node_hash_bytes<W: Write>(&mut self, ptr: &TriePtr, w: &mut W) -> Result<(), Error> {
-        let trie_offset = self.file.get_trie_offset(self.db, self.block_id)?;
-        self.file
-            .seek(SeekFrom::Start(trie_offset + (ptr.ptr() as u64)))?;
-        let hash_buff = read_hash_bytes(self.file)?;
-        w.write_all(&hash_buff).map_err(|e| e.into())
+        Ok(compressed_bytes)
     }
-}
 
-impl TrieFile {
-    /// Determine the file offset in the TrieFile where a serialized trie starts.
-    /// The offsets are stored in the given DB, and are cached indefinitely once loaded.
-    pub fn get_trie_offset(&mut self, db: &Connection, block_id: u32) -> Result<u64, Error> {
-        let offset_opt = match self {
-            TrieFile::RAM(ref ram) => ram.trie_offsets.get(&block_id),
-            TrieFile::Disk(ref disk) => disk.trie_offsets.get(&block_id),
-        };
-        match offset_opt {
-            Some(offset) => Ok(*offset),
-            None => {
-                let (offset, _length) = trie_sql::get_external_trie_offset_length(db, block_id)?;
-                match self {
-                    TrieFile::RAM(ref mut ram) => ram.trie_offsets.insert(block_id, offset),
-                    TrieFile::Disk(ref mut disk) => disk.trie_offsets.insert(block_id, offset),
-                };
-                Ok(offset)
-            }
+    /// Append a serialized and compressed trie to the TrieFile.
+    /// Returns the offset at which it was appended.
+    pub fn append_trie_blob(
+        &mut self,
+        compression_type: &BlobCompressionType,
+        db: &Connection,
+        buf: &[u8],
+        offset_override: Option<u64>,
+    ) -> Result<BlobStorageResult, Error> {
+        let offset: u64;
+        if offset_override.is_some() {
+            offset = offset_override.unwrap();
+        } else {
+            offset = trie_sql::get_external_blobs_length(db)?;
         }
+        self.seek(SeekFrom::Start(offset))?;
+
+        let mut compression_elapsed: Duration = Duration::from_micros(0);
+        let compression_bench = SystemTime::now();
+
+        let storage_bytes = match compression_type {
+            BlobCompressionType::None => buf.into(),
+            _ => {
+                let compressed = Self::compress_blob(compression_type, buf)?;
+                compression_elapsed = compression_bench.elapsed().unwrap();
+                compressed
+            }
+        };
+
+        let storage_bytes_slice = storage_bytes.as_slice();
+
+        eprintln!("Write trie of {} (input) and {} (to-disk) bytes at {} using compression type {}. Compression time {:?}", 
+            buf.len(),
+            storage_bytes.len(),
+            offset,
+            &compression_type,
+            compression_elapsed);
+
+        self.seek(SeekFrom::Start(offset))?;
+        self.write_all(storage_bytes_slice)?;
+        self.flush()?;
+        //self.sync_data()?;
```

Comment:
`flush()` by itself does not persist data.  In the case of writing data to a file, you must also call `.sync_data()` on the file descriptor to ensure that all newly-written blocks are persisted.